File: 181004_python_speed/mysum.c
# include <stdio.h>

/* return the sum of the elements of data
 */
double mysum(double data[], int m) {
    double result = 0;
    for (int i = 0; i < m; i++) {
        result += data[i];
    }
    return result;
}

int main() {
    int m = 100000;
    int n = 10000;

    // create an array of floats from 0 to m
    double data[m];
    for (int i = 0; i < m; i++) {
        data[i] = (double) i;
    }

    // compute the sum of the array n times
    double result = 0;
    for (int i = 0; i < n; i++) {
        result += mysum(data, m);
    }

    // print summary and exit
    result /= m * n;
    printf("computed the sum %d floats %d times. normalized result=%f\n",
           m, n, result);
    return 0;
}

File: 181004_python_speed/mysum.py
def mysum(data):
    '''return the sum of the elements of data'''
    result = 0
    for value in data:
        result += value
    return result

def main(m=100000, n=10000):

    # create a list of floats from 0 to m
    data = [float(i) for i in range(m)]

    # compute the sum of the list n times
    result = 0.0
    for _ in range(n):
        result += mysum(data)

    # print summary and exit
    result /= m * n
    print(f'computed the sum of {m} floats {n} times. normalized result={result}')
    return

if __name__ == '__main__':
    main()
